### DISCUSSION TITLE
static absorption time

### DISCUSSION THREAD

**Discussion by mdkeil** [Louisvdw/dbus-serialbattery#807] [Q&A]
Is it possible to set a static absorption time (e.g. 30 min) or an minimum time before the driver goes to "float"?

Background:
Seplos BMS 10E requires 30min on 16x Cell OVA to set the SOC to 100%.

Edit: with NEEY active balancer at 2A It takes less than 15 minutes for the cells to be balanced (deviation 5mV)

**Comment by mr-manuel** [comment]
Yes, just use

https://github.com/Louisvdw/dbus-serialbattery/blob/88274bf54eb21733d0b77008d4453a6297e6e827/etc/dbus-serialbattery/config.default.ini#L103-L107

but with linear mode the BMS goes to float only, if the cells are balanced. Therefore no static time needed.

https://github.com/Louisvdw/dbus-serialbattery/blob/88274bf54eb21733d0b77008d4453a6297e6e827/etc/dbus-serialbattery/config.default.ini#L95-L101

**Comment by mdkeil** [comment]
Thxs for the answer; I know these options- but how I can only use the step mode? or deactivate the linear mode?

**Reply by mr-manuel** [reply]
By reading the `config.sample.ini` and changing the corresponding value in your `config.ini`.

https://github.com/Louisvdw/dbus-serialbattery/blob/88274bf54eb21733d0b77008d4453a6297e6e827/etc/dbus-serialbattery/config.default.ini#L56-L62

**Reply by mdkeil** [reply]
..but I want to use step mode only for CVL.

**Reply by mr-manuel** [reply]
Sorry but I can't understand what you really want.

> Is it possible to set a static absorption time (e.g. 30 min) or an minimum time before the driver goes to "float"?

This is possible by setting `MAX_VOLTAGE_TIME_SEC = 1800` indipendent of linear or step mode. It won't work, if you are using an old version. Install the `nightly` build from the `dev` branch.

**Reply by mdkeil** [reply]
> with NEEY active balancer at 2A It takes less than 15 minutes for the cells to be balanced (deviation 5mV)

..at max voltage


Thats my "problem".. with linear mode:  float voltage will be set after 10-15minutes.. but I need the max voltage for minimum 30min for SEPLOS BMS to reset the SOC to 100%.

If I set LINEAR_LIMITATION_ENABLE to false, I will lost the linear mode for CCL and DCL too.. or I don't realy understand the descriptions in the sample-file.

**Reply by mr-manuel** [reply]
The only problem is probably a language barrier. Just set `MAX_VOLTAGE_TIME_SEC = 1800`. This will keep max voltage at least for 1800 seconds, which are 30 minutes or 0.5 hours. Indipendent of step or linear mode.

**Reply by mdkeil** [reply]
> ;     Linear mode: If cells are balanced keep max voltage for further MAX_VOLTAGE_TIME_SEC seconds

ok that fact i did not read correctly.. ("for **further**") Therfore MAX_VOLTAGE_TIME_SEC = 1800 should be the right settings for me.

**Reply by mdkeil** [reply]
it doesnt work..

with `LINEAR_LIMITATION_ENABLE = True`  and `MAX_VOLTAGE_TIME_SEC = 1800`

because of setting to float mode 300s after cell diff is below CELL_VOLTAGE_DIFF_KEEP_MAX_VOLTAGE_UNTIL = 0.010

**Reply by mdkeil** [reply]
some screenshots from my monitoring:

![absorp_1](https://github.com/Louisvdw/dbus-serialbattery/assets/8126652/9647c99d-ef79-41b0-8c46-3ac5ea1d551d)
![absorp_2](https://github.com/Louisvdw/dbus-serialbattery/assets/8126652/86e38d70-a115-43de-9584-6e6b0425b34b)
![absorp_3](https://github.com/Louisvdw/dbus-serialbattery/assets/8126652/cc2977a9-4ec8-49d8-a236-fcafb57a0709)

max voltage: 55,6V

15:41:16 - max voltage at 55,6V
15:49:01 - cell diff voltage below 10mv
15:54:11 - start to drop down voltage

```
; Specify how long the max voltage should be kept 
 ;     Step mode: If reached then switch to float voltage 
 ;     Linear mode: If cells are balanced keep max voltage for further MAX_VOLTAGE_TIME_SEC seconds 
 MAX_VOLTAGE_TIME_SEC = 900 
```

example from Charge Voltage limitation

```
The battery reached max voltage of 55.2V and the max cell difference is 0.010V, then switch to float
;          voltage of 53.6V after 300 additional seconds to don't stress the batteries. [..]
```

The implemantation does not fit with the discription.

**Reply by mdkeil** [reply]
updated from last release (from may) to last nightly@dev, further testing tomorrow.

**Reply by mr-manuel** [reply]
Please put a hashtag `#` in the front of this line

https://github.com/Louisvdw/dbus-serialbattery/blob/4c2b89ca11bbc343a77f67b2c6dcb0e663f4157d/etc/dbus-serialbattery/battery.py#L443

and reinstall the driver by running `/data/etc/dbus-serialbattery/reinstall-local.sh`. This allows you to see extended debugging info unter the parameter page. There you also see when the timer for `MAX_VOLTAGE_TIME_SEC` starts.

**Comment by mdkeil** [comment]
not enought sun-power to get the battery full of charge, but the driver shows "Absorption (Linear Mode)" as charge mode all the time.

![absorp_4](https://github.com/Louisvdw/dbus-serialbattery/assets/8126652/ab6ea0eb-5560-46c4-ac24-6bbb9eaf98cc)
![absorp_5](https://github.com/Louisvdw/dbus-serialbattery/assets/8126652/a5a34eb4-6254-4e90-9763-3eb06b0057e3)

**Reply by mr-manuel** [reply]
You are answering to your question already yourself. How should the charger switch to float, if there is not enough sun to complete bulk/absorption?

**Reply by mdkeil** [reply]
before switching to nightly-build the charge mode was a little bit more "detailed" with different states starting from "Bulk (Linear Mode)". 
So I was surprised that it only ever shows "Absorption" (before Float state)
